sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/ReadFromKafkaDoFn.java

```diff
@@ -424,6 +425,29 @@ public ProcessContinuation processElement(
     }
   }
 
+  // see https://github.com/apache/beam/issues/25962
+  private ConsumerRecords<byte[], byte[]> poll(
+      Consumer<byte[], byte[]> consumer, TopicPartition topicPartition) {
+    final Stopwatch sw = Stopwatch.createStarted();
+    long previousPosition = -1;
+    while (true) {
+      final ConsumerRecords<byte[], byte[]> rawRecords =
+          consumer.poll(KAFKA_POLL_TIMEOUT.minus(sw.elapsed()));
+      if (!rawRecords.isEmpty()) {
+        // return as we have found some entries
+        return rawRecords;
+      }
+      if (previousPosition == (previousPosition = consumer.position(topicPartition))) {
```

Comment:
I'm not completely sure we need this. There is a pretty high chance it works flawlessly even without this.

Reply:
(end of stream -> no offset change possible)
pre-3.2.0+EOS -> poll should block for timeout already, return empty, *both position and timeout check should return*
post-3.2.0+EOS -> poll should block for timeout already, return empty, *both position and timeout check should return*

(small gap -> timeout is enough to iterate over it)
pre-3.2.0+small gap -> poll blocks until it returns a non-empty result, *non-empty check should return*
post-3.2.0+small gap -> poll return empty immediately, *every check fails*

(big gap -> timeout isn't enough to iterate over it)
pre-3.2.0+big gap -> poll should block for timeout already, return empty, *timeout check should return*
post-3.2.0+big gap -> poll return empty immediately, *every check fails*

So _theoretically_ there isn't any scenario where the position check triggers the return, but nothing else.